bin/oneoff/backfill_data/backfill_script_ids_channel_tokens_1.rb

```diff
@@ -0,0 +1,204 @@
+      if script_id_by_parent_user_levels
+        channel_token.update_attributes(script_id: script_id_by_parent_user_levels) unless is_dry_run
+        backfill_count += 1
+        next
+      end
+    end
+
+    # if the level is associated with only one script_level, use that script
+    if associated_script_levels.length == 1
+      script_id = associated_script_levels[0].script_id
+      channel_token.update_attributes(script_id: script_id) unless is_dry_run
+      backfill_count += 1
+      next
+    end
+
+    potential_associated_scripts = associated_script_levels.map(&:script_id)
+
+    # if the level is associated with only one script_level where the script is visible,
+    # use that script
+    if visible_script_id = script_id_by_visible_scripts(potential_associated_scripts)
+      channel_token.update_attributes(script_id: visible_script_id) unless is_dry_run
+      backfill_count += 1
+      next
+    end
+
+    # if the user is in or owns a section where the script assigned matches a potential script
+    # that could be associated with the level, use that script
+    if user_id.present? && section_script_id = script_id_by_section(user_id, potential_associated_scripts)
+      channel_token.update_attributes(script_id: section_script_id) unless is_dry_run
+      backfill_count += 1
+      next
+    end
+
+    levels_missing_backfills.push(level.id)
+    unable_to_backfill += 1
+  end
+
+  puts "backfilled #{backfill_count} script ids, unable to backfill script id for #{unable_to_backfill} channel tokens"
+
+  # for investigation purposes
+  puts "unfilled levels:"
+  puts levels_missing_backfills.uniq!
+end
+
+def get_parent_level_ids(level)
+  parent_level_child_levels = level.levels_parent_levels
+  return unless parent_level_child_levels.any?
+
+  parent_level_child_levels.map(&:parent_level_id)
+end
+
+def get_script_id_by_parent_levels(parent_level_ids, user_id)
+  parent_level_ids.each do |level_id|
+    script_id = script_id_by_user_level(user_id, level_id)
+    return script_id if script_id
+  end
+end
+
+# Given the script_levels that a channel token may be associated with, this method
+# returns the script_id for the associated script if there is only one publicly visible script
+def script_id_by_visible_scripts(associated_script_ids)
+  visible_published_states = [
+    SharedConstants::PUBLISHED_STATE.stable,
+    SharedConstants::PUBLISHED_STATE.preview,
+    SharedConstants::PUBLISHED_STATE.beta,
+    SharedConstants::PUBLISHED_STATE.in_development
+  ]
+
+  visible_associated_scripts = Script.where(
+    id: associated_script_ids,
+    published_state: visible_published_states
+  )
+
+  return visible_associated_scripts[0].id if visible_associated_scripts.count == 1
+end
+
+# Given a channel token, this method returns the associated script_id if it can
+# be identified based on user_level
+def script_id_by_user_level(user_id, level_id)
```

Comment:
(minor) I was a little confused by this name at first (probably because "by" signifies a map to me, coming from a java background).  Would `script_id_from_user_level` be a good alternative?